### PR TITLE
Re-worked patch engine.

### PR DESCRIPTION
**Pull request by kyenos**
A number of edge cases, bugs and just flat out assumptions caused me to re-think the underlying reflection code and ultimately justified a refactoring.

One of the things I took liberty with, is the information kept in the change log, I've added a copy of an items parent to satisfy the edge case of a target that is complex or a slice and doesn't contain the original 'untouched' items necessary to make the match. I recognize that in complex situations, this could significantly increase the memory footprint of a change log unnecessarily and have added an option to avoid this allocation if the user will not find themselves needing it by default.  This caused me to make some rather invasive changes to diff itself which may be unwanted upstream. Im open to critical feedback on this note.

Since it's a refactor, a lot of changes are inside... including:

- Changes to the change log structure to include a reference to a parent struct, should there be one. This satisfies an edge case of the target being an unallocated complex struct. This comes a a complexity expense as well as a memory footprint cost. It does not add a computational expense to the operation.
- An option to turn off the above edge case feature for memory constrained environments
- Full recursion into complex structures. Note that this is some what opinionated but should operate intuitively
- Support for complex map keys. This comes with the need for type information in the change path. In order to support this, struct keys are marshaled into their equivalent strings using msgpack. I'm considering the ability to turn this off however no option is there yet.
- I've split out the code to mimic the layout of the diff_ files for consistency
- Updated tests to account for the differing issues and new code
- Added the ability to pass Differ options directly to the function Diff when making the call.
- Corrected the lack of honoring all options when setting values
- Updated the README to include patch information and examples

### CODE REVIEW

**Comment by purehyperbole** [comment]
Hey @kyenos thank you for taking the time to update the diff functionality. At a cursory glance, the changes look really great!

Given the size of the PR and extent of the changes, it may take me a bit of time to properly go through everything (hopefully I can find some time before the end of this this week to go through everything).

File: diff.go [hunk 1]
```diff
@@ -238,7 +254,11 @@ func idstring(v interface{}) string {
 	case int:
 		return strconv.Itoa(v.(int))
 	default:
-		return fmt.Sprint(v)
+		b, err := msgpack.Marshal(v)
```
Comment (purehyperbole) [inline diff.go:257]: For the sake of preserving the current output format for other users, I think we need to hide this behind an option to the differ

**Review by purehyperbole — CHANGES_REQUESTED** [review]

**Comment by purehyperbole** [comment]
Hey @kyenos 

The changes look awesome! Only one suggested change (which you also mention in your description). 

Once that's in place, I'm happy to merge this.

Thanks again!

File: diff.go (hunk 1 above)
Comment (kyenos) [reply diff.go:259]: Honestly, I sort of saw this one coming. It was a trade off between keeping backwards compatibility and keeping the output of diff compatible with patch by default.

I did / do however agree with the ask and think we should protect downstream compatibility, breaking users existing code will probably have them walking away.

I put the marshaling behind the StructMapKeySupport option. This however increased complexity enough to drop the unit tests below 80% so Ive done some additional due-diligence and  shored up the tests by adding some coverage for uint (which was the only set of code at 0%)  bringing us back up to 80% coverage across the file set.

**Comment by purehyperbole** [comment]
Hey @kyenos 

Many thanks for updating the PR and again for the contribution. I'll get this merged and cut a release!

**Pull request merged**